### PR TITLE
[SYCLomatic] Add new option --gen-helper-function

### PR DESCRIPTION
**Pull request by zhiweij1**
Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

**Review by ziranzha — APPROVED** [review]

File: clang/include/clang/DPCT/DPCTOptions.inc [hunk 1]
```diff
@@ -434,4 +434,8 @@ DPCT_NON_ENUM_OPTION(
         "Limit extension change to .cu and .cuh files only. Default: off."),
     llvm::cl::value_desc("file"), llvm::cl::cat(DPCTCat), llvm::cl::init(false))
 
+DPCT_NON_ENUM_OPTION(DPCT_OPT_TYPE(static llvm::cl::opt<bool>), GenHeadersInOutRoot, "gen-headers-in-out-root",
```
Comment (zhimingwang36) [inline clang/include/clang/DPCT/DPCTOptions.inc]:
```suggestion
DPCT_NON_ENUM_OPTION(DPCT_OPT_TYPE(static llvm::cl::opt<bool>), GenHeadersInOutRoot, "gen-helper-function",
```

File: clang/include/clang/DPCT/DPCTOptions.inc [hunk 2]
```diff
@@ -434,4 +434,8 @@ DPCT_NON_ENUM_OPTION(
...
+        llvm::cl::desc("Generate all helper header files in the --out-root directory. Default: off."),
```
Comment (zhimingwang36) [inline clang/include/clang/DPCT/DPCTOptions.inc]: ### 
```suggestion
        llvm::cl::desc("Generate all helper function files in the --out-root directory. Default: off."),
```

File: clang/include/clang/DPCT/DPCTOptions.inc [hunk 3]
```diff
@@ -434,4 +434,8 @@ DPCT_NON_ENUM_OPTION(
...
+        llvm::cl::cat(DPCTCat), llvm::cl::init(false))
+
```
Comment (zhimingwang36) [inline clang/include/clang/DPCT/DPCTOptions.inc:440]: let's make this option can work standalone.

File: clang/include/clang/DPCT/DPCTOptions.inc [hunk 4]
```diff
@@ -434,8 +434,8 @@ DPCT_NON_ENUM_OPTION(
         "Limit extension change to .cu and .cuh files only. Default: off."),
     llvm::cl::value_desc("file"), llvm::cl::cat(DPCTCat), llvm::cl::init(false))
 
-DPCT_NON_ENUM_OPTION(DPCT_OPT_TYPE(static llvm::cl::opt<bool>), GenHeadersInOutRoot, "gen-headers-in-out-root",
-        llvm::cl::desc("Generate all helper header files in the --out-root directory. Default: off."),
+DPCT_NON_ENUM_OPTION(DPCT_OPT_TYPE(static llvm::cl::opt<bool>), GenHelperFunction, "gen-helper-function",
+        llvm::cl::desc("Generate all helper function files in the --out-root directory. Default: off."),
```
Comment (zhimingwang36) [inline clang/include/clang/DPCT/DPCTOptions.inc]:
```suggestion
        llvm::cl::desc("Generate helper function files in the --out-root directory. Default: off."),
```

File: clang/lib/DPCT/GenHelperFunction.cpp [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+  }
+}
+
+void replaceEndOfLine(std::string &StrNeedProcess) {
+#ifdef _WIN64
+  replaceAllOccurredStrsInStr(StrNeedProcess, "\n", "\r\n");
+#endif
+}
+
+std::unordered_map<std::string, std::string> HelperFileNameMap{
```
Comment (zhimingwang36) [inline clang/lib/DPCT/GenHelperFunction.cpp:120]: try to reuse HelperFileNames.inc

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**